### ISSUE TITLE
Tearing flip support

### ISSUE THREAD

**Issue by YaLTeR** [YaLTeR/niri#844]
Useful for games and stuff to get the lowest input latency.

We're mainly waiting for https://github.com/Smithay/smithay/pull/1335 which is waiting for some driver stuff (?). Then on the niri side we can make tearing a window rule just like [on-demand VRR](https://github.com/YaLTeR/niri/wiki/Configuration:-Window-Rules#variable-refresh-rate).

There's also https://github.com/Smithay/smithay/pull/1325 for the Wayland protocol, which is needed so windows can automatically request tearing without a manual window rule.

**Comment by witchlliee** [comment]
Any update on this one? I migrated to niri and i'm loving it but i sure miss the tearing protocol.

**Comment by shibahaczix** [comment]
Yeah, I also migrated back to niri, it would be cool if we had tearing support for da gamerz!

**Comment by exodist** [comment]
Just migrated, and I also would like this.

**Comment by shibahaczix** [comment]
https://github.com/Smithay/smithay/pull/1335 kinda annoying that this isn't higher priority. 

Not everyone has VRR.

**Comment by YaLTeR** [comment]
If you're a paying system76 customer then I suppose there might be some way for you to inform them of what should be "higher priority" in cosmic, and by extension Smithay

**Comment by Gapva** [comment]
this seriously needs to be higher priority

**Comment by JuxGD** [comment]
Would it be silly to keep a fork in the meantime?

**Comment by Sempyos** [comment]
> Would it be silly to keep a fork in the meantime?

I feel some people may appreciate that if you do

**Comment by shibahaczix** [comment]
> Would it be silly to keep a fork in the meantime?

Please do!

**Comment by urayde** [comment]
> > Would it be silly to keep a fork in the meantime?
> 
> Please do!

You could try mine for now then :>

**Comment by shibahaczix** [comment]
<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/9c3c4901-a45f-4939-ad64-27b3474b56a6" />

Seems to be working fine?
Thought I feel a slight jitter in my input idk
@urayde

**Comment by JuxGD** [comment]
> > Would it be silly to keep a fork in the meantime?
> 
> I feel some people may appreciate that if you do

I wasn't offering lol Idk how to do that

**Comment by shibahaczix** [comment]
OK I know why, if you set a fps limit the frametimes cause input stutters on immediate mode

**Comment by shibahaczix** [comment]
Idk if this is placebo but the latency feels wayyy better 
my mouse feels like im sliding on wet ice

**Comment by shibahaczix** [comment]
if someone wants to test this on nixos do this
```nix
{
  inputs = {
    niri = {
      url = "github:sodiboo/niri-flake";
      inputs.niri-unstable.url = "github:urayde/niri";
      inputs.nixpkgs.follows = "nixpkgs";
    };
  }
}
```

**Comment by JuxGD** [comment]
sodiboo my goat

**Comment by akmalm007** [comment]
@urayde Thank you for the Fork, I feeling the same thing as @shibahaczix mention the latency is much better than in KDE and I got constant FPS in Left 4 Dead 2. 

Looking forward for the next update to support tearing

**Comment by Kreevoz** [comment]
After spotting the fork I've finally switched to Niri too. Thank you for the nix config snippet, @shibahaczix 

I feel Niri by itself is already fairly speedy in terms of latency, but tearing is extremely useful when using a wacom pen to do illustrations/painting/concepting. It lowers the latency between pen input and brush response on the canvas very noticably. 

Have not noticed any bugs or unexpected behavior so really @urayde 's fork gets a +1 from me with the plea to the Niri developer that this feature gets picked up and merged into main. It really works and provides practical value.

**Comment by JuxGD** [comment]
Same here, no problems with tearing as far as I know. Merging would mean depending on an unofficial Smithay fork though, I think trying to get upstream Smithay to implement tearing flip support is better

Worth mentioning that Smithay/smithay#1335, which this issue was waiting on, is apparently outdated, or at least not the only one :o There's a new PR @ Smithay/smithay#1948 that's not a draft and is waiting for approval. @YaLTeR mentioning you so you know to pay attention to that PR

I'm gonna go bump that PR now lol

**Comment by ghost** [comment]
Would bringing tearing to Niri also bring explicit sync (like @urayde 's fork) or are they planned on being brought in separately? Also, the fork works great by the way!
Edit: I have noticed that clients where tearing is enabled (in the fork) seem to be getting frame callbacks instantly, busy looping the CPU. I don't know if this is intended behavior, but they seem like they should be separated. The fork also seems to have higher idle CPU usage even when tearing is disabled, so maybe these are things to work on before adding tearing to the main Niri repo (or maybe they are issues with the changes to Smithay needed to add tearing).

**Comment by JuxGD** [comment]
> I have noticed that clients where tearing is enabled (in the fork) seem to be getting frame callbacks instantly, busy looping the CPU. I don't know if this is intended behavior, but they seem like they should be separated. The fork also seems to have higher idle CPU usage even when tearing is disabled, so maybe these are things to work on before adding tearing to the main Niri repo (or maybe they are issues with the changes to Smithay needed to add tearing).

I wonder if the PRs on upstream have this problem too

**Comment by ghost** [comment]
> I wonder if the PRs on upstream have this problem too

Yeah, I was thinking that that was a possibility, but the CPU usage stays higher than the official Niri package even when no GPU accelerated clients are running. Also when tearing is disabled, the only difference should be the implementaiotn of DRM syncobj which Cosmic has, and I would assume that would be a known problem if it were a Smithay issue (unless it is an issue with the Smithay fork).

The actual CPU usage on regular Niri on idle for me is 0-1% and with the tearing fork it is ~3%. Small difference, but it is extraordinarily consistent, and I can confidently say it is not error.

As for the frame callback issue, it only happens when tearing is enabled, could be the Smithay fork or Niri fork.

I should say that I know almost nothing about Rust, DRM/KMS, or Smithay; these are just some things that I've noticed and my guesses (which should be taken with a grain of salt) as to what could be the cause.

**Comment by JuxGD** [comment]
Hm, CPU usage for me is like 0.3% with no processes other than systemd with its services and waybar, awww, etc

**Comment by ghost** [comment]
Myabe its a quirk with my drivers. Im on an Intel 13th gen mobile chip (with updated microcode b/c i know 13/14 gen Intel had bad microcode) with an iGPU. Also for me, same things running, waybar, awww, systemd, etc.

**Comment by urayde** [comment]
> Would bringing tearing to Niri also bring explicit sync (like [@urayde](https://github.com/urayde) 's fork) or are they planned on being brought in separately?

Those are not really related, it's just the stuff I wanted to try to work on.

**Comment by ghost** [comment]
> Those are not really related, it's just the stuff I wanted to try to work on.

Yeah, i understand; i think syncobj would be nice to have, thats why i asked. You did a great job with the fork though, it seems my hardware is just odd. The frame callback thing is probably worth looking into (unless it's intentional), but this is probably not the place to discuss that.

Edit: using the package from the AUR fixed the CPU usage issue. When I built from source I built in release mode, so I don't know what the cause was.

**Comment by ItsLiyua** [comment]
I made a fork of sordiboo's niri-flake that has support for the `allow-tearing` windowrule.
https://github.com/ItsLiyua/niri-flake
Just add something along the lines of
```nix
window-rules = [
  {
    matches = [ { title = "Minecraft.*"; } ];
    allow-tearing = true;
  }
];
```
to your nix config.

I did not add support for the debug option to force tearing globally as it was not relevant to me but if that's necessary I (or you) can add it as well.
